ssl.c

```diff
@@ -92,142 +92,172 @@ void
+
+	return cert;
 }
 
 void
 rdssl_cert_free(RDSSL_CERT * cert)
 {
-	X509_free(cert);
+	gnutls_free(cert);
 }
 
+
+/*
+ * AFAIK, there's no way to alter the decoded certificate using GnuTLS.
+ *
+ * Upon detecting "problem" (wrong public RSA key OID) certificate
+ * we basically have two options:
+ *
+ * 1)) encode certificate back to DER, then parse it using libtasn1,
+ * fix public key OID (set it to 1.2.840.113549.1.1.1), encode to DER again
+ * and finally reparse using GnuTLS
+ *
+ * 2) encode cert back to DER, get RSA public key parameters using libtasn1
+ *
+ * Or can rewrite the whole certificate related stuff later.
+ */
+
 /* returns newly allocated RDSSL_RKEY or NULL */
 RDSSL_RKEY *
 rdssl_cert_to_rkey(RDSSL_CERT * cert, uint32 * key_len)
 {
-	EVP_PKEY *epk = NULL;
-	RDSSL_RKEY *lkey;
-	int nid;
 	int ret;
+	RDSSL_RKEY *pkey;
+	gnutls_datum_t m, e;
 
-	const unsigned char *p;
-	int pklen;
-
-	RSA *rsa = NULL;
+	unsigned int algo, bits;
+	char oid[64];
+	size_t oid_size = sizeof(oid);
 
-	/* By some reason, Microsoft sets the OID of the Public RSA key to
-	   the oid for "MD5 with RSA Encryption" instead of "RSA Encryption"
+	uint8_t data[2048];
+	size_t len;
 
-	   Kudos to Richard Levitte for the following (. intiutive .) 
-	   lines of code that resets the OID and let's us extract the key. */
+	algo = gnutls_x509_crt_get_pk_algorithm(*cert, &bits);
 
-	X509_PUBKEY *key = NULL;
-	X509_ALGOR *algor = NULL;
+	if (algo == GNUTLS_PK_RSA) {
 
-	key = X509_get_X509_PUBKEY(cert);
-	if (key == NULL)
-	{
-		logger(Protocol, Error,
-		       "rdssl_cert_to_key(), failed to get public key from certificate");
-		rdssl_log_ssl_errors("rdssl_cert_to_key()");
+		if ((ret = gnutls_x509_crt_get_pk_rsa_raw(*cert, &m, &e)) !=  GNUTLS_E_SUCCESS) {
+			logger(Protocol, Error, "%s:%s:%d: Failed to get RSA public key parameters from certificate. GnuTLS error = 0x%02x (%s)\n",
+					__FILE__, __func__, __LINE__, ret, gnutls_strerror(ret));
+			return NULL;
+		}
 
-		return NULL;
-	}
+	} else if (algo == GNUTLS_E_UNIMPLEMENTED_FEATURE) {
 
-	ret = X509_PUBKEY_get0_param(NULL, NULL, 0, &algor, key);
-	if (ret != 1)
-	{
-		logger(Protocol, Error,
-		       "rdssl_cert_to_key(), failed to get algorithm used for public key");
-		rdssl_log_ssl_errors("rdssl_cert_to_key()");
+		// manpage says that this function is useful when gnutls_x509_crt_get_pk_algorithm() returns GNUTLS_PK_UNKNOWN
+		if ((ret = gnutls_x509_crt_get_pk_oid(*cert, oid, &oid_size)) != GNUTLS_E_SUCCESS) {
+			logger(Protocol, Error, "%s:%s:%d: Failed to get OID of public key algorithm. GnuTLS error = 0x%02x (%s)\n",
+					__FILE__, __func__, __LINE__, ret, gnutls_strerror(ret));
+			return NULL;
+		}
 
-		return NULL;
-	}
+		/* By some reason, Microsoft sets the OID of the Public RSA key to
+		   the oid for "MD5 with RSA Encryption" instead of "RSA Encryption"
 
-	nid = OBJ_obj2nid(algor->algorithm);
+		   Kudos to Richard Levitte for the finding this and proposed the fix
+		   using OpenSSL. */
```

Comment:
I've left it as token of gratitude to the person who first found it and fixed.
Sure it could be removed but I'd prefer to leave it.
Anyway, I'm open to suggestions:)